API Core PyType fails intermittently/changing errors

https://github.com/googleapis/google-cloud-python/pull/6769

Copying from #6769 and @rchen152 


> @crwilcox It looks like what's happening is that pytype is somehow finding multiple files named path_template.py (and timeout.py and grpc_helpers.py - the error probably differs because it varies which file pytype discovers first). The other thing I noticed is that the file count in the status message ("Analyzing 53 sources with 0 dependencies") differs from the one I see (28 sources). Any idea why there might be two copies of some source files?
> 
> EDIT: if there are two files with the same name in different directories, this may be google/pytype#198.

When I edit `setup.cfg` to include only `google/`, then I see the following error:

```bash
$ nox -re pytype
nox > Running session pytype
nox > Re-using existing virtualenv at .../api_core/.nox/pytype.
nox > pip install --upgrade . grpcio >= 1.8.2 grpcio-gcp >= 0.2.2 pytype >= 2018.9.26
nox > pytype 
Analyzing 26 sources with 0 dependencies
ninja: Entering directory `.../api_core/pytype_output'
[3/26] pytype-single --imports_info /h..._core/google/api_core/path_template.py
FAILED: .../api_core/pytype_output/pyi/google/api_core/iam.pyi 
pytype-single --imports_info .../api_core/pytype_output/imports/google.api_core.iam.imports --module-name google.api_core.iam -V 3.6 -o .../api_core/pytype_output/pyi/google/api_core/iam.pyi --analyze-annotated --nofail --quick .../api_core/google/api_core/iam.py
File ".../api_core/google/api_core/iam.py", line 248, in <module>: No attribute 'register' on Type[MutableMapping] [attribute-error]

For more details, see https://github.com/google/pytype/blob/master/docs/errors.md#attribute-error.
[26/26] pytype-single --imports_info /...gle/api_core/operations_v1/__init__.py
ninja: build stopped: cannot make progress due to previous errors.
nox > Command pytype  failed with exit code 1
nox > Session pytype failed.
```

I don't understand the error: `collections.abc.MutableMapping` clearly has a `register` method:  it [calls it on the `dict` type in its own source code](https://github.com/python/cpython/blob/1133a8c0efabf6b33a169039cf6e2e03bfe6cfe3/Lib/_collections_abc.py#L859).

--- a/.gitignore
+++ b/.gitignore
@@ -69,3 +69,4 @@ generated_python/
 cloud-bigtable-client/
 googleapis-pb/
 grpc_python_venv/
+pytype_output/

--- a/api_core/google/api_core/iam.py
+++ b/api_core/google/api_core/iam.py
@@ -243,6 +243,3 @@ class Policy(collections_abc.MutableMapping):
                 del resource["bindings"]
 
         return resource
-
-
-collections_abc.MutableMapping.register(Policy)

--- a/api_core/setup.cfg
+++ b/api_core/setup.cfg
@@ -4,6 +4,6 @@ universal = 1
 [pytype]
 python_version = 3.6
 inputs =
-    .
+    google/
 exclude =
     tests/

--- a/api_core/tests/unit/test_iam.py
+++ b/api_core/tests/unit/test_iam.py
@@ -90,9 +90,12 @@ class TestPolicy:
         MEMBER = "user:phred@example.com"
         expected = set([MEMBER])
         policy = self._make_one()
-        with warnings.catch_warnings():
-            warnings.simplefilter("always")
+
+        with warnings.catch_warnings(record=True) as warned:
             policy.owners = [MEMBER]
+
+        warning, = warned
+        assert warning.category is DeprecationWarning
         assert policy[OWNER_ROLE] == expected
 
     def test_editors_getter(self):
@@ -111,9 +114,12 @@ class TestPolicy:
         MEMBER = "user:phred@example.com"
         expected = set([MEMBER])
         policy = self._make_one()
-        with warnings.catch_warnings():
-            warnings.simplefilter("always")
+
+        with warnings.catch_warnings(record=True) as warned:
             policy.editors = [MEMBER]
+
+        warning, = warned
+        assert warning.category is DeprecationWarning
         assert policy[EDITOR_ROLE] == expected
 
     def test_viewers_getter(self):
@@ -132,9 +138,12 @@ class TestPolicy:
         MEMBER = "user:phred@example.com"
         expected = set([MEMBER])
         policy = self._make_one()
-        with warnings.catch_warnings():
-            warnings.simplefilter("always")
+
+        with warnings.catch_warnings(record=True) as warned:
             policy.viewers = [MEMBER]
+
+        warning, = warned
+        assert warning.category is DeprecationWarning
         assert policy[VIEWER_ROLE] == expected
 
     def test_user(self):
@@ -240,17 +249,20 @@ class TestPolicy:
         assert policy.to_api_repr() == {}
 
     def test_to_api_repr_binding_w_duplicates(self):
+        import warnings
         from google.api_core.iam import OWNER_ROLE
 
         OWNER = "group:cloud-logs@google.com"
         policy = self._make_one()
-        policy.owners = [OWNER, OWNER]
+        with warnings.catch_warnings(record=True):
+            policy.owners = [OWNER, OWNER]
         assert policy.to_api_repr() == {
             "bindings": [{"role": OWNER_ROLE, "members": [OWNER]}]
         }
 
     def test_to_api_repr_full(self):
         import operator
+        import warnings
         from google.api_core.iam import OWNER_ROLE, EDITOR_ROLE, VIEWER_ROLE
 
         OWNER1 = "group:cloud-logs@google.com"
@@ -265,9 +277,10 @@ class TestPolicy:
             {"role": VIEWER_ROLE, "members": [VIEWER1, VIEWER2]},
         ]
         policy = self._make_one("DEADBEEF", 17)
-        policy.owners = [OWNER1, OWNER2]
-        policy.editors = [EDITOR1, EDITOR2]
-        policy.viewers = [VIEWER1, VIEWER2]
+        with warnings.catch_warnings(record=True):
+            policy.owners = [OWNER1, OWNER2]
+            policy.editors = [EDITOR1, EDITOR2]
+            policy.viewers = [VIEWER1, VIEWER2]
         resource = policy.to_api_repr()
         assert resource["etag"] == "DEADBEEF"
         assert resource["version"] == 17

